mempool/cat/reactor.go

```diff
@@ -0,0 +1,443 @@
+func (memR *Reactor) OnStart() error {
+	if memR.opts.ListenOnly {
+		memR.Logger.Info("Tx broadcasting is disabled")
+		return nil
+	}
+	go func() {
+		for {
+			select {
+			case <-memR.Quit():
+				return
+
+			// listen in for any newly verified tx via RFC, then immediately
+			// broadcasts it to all connected peers.
+			case nextTxKey := <-memR.mempool.next():
+				wtx := memR.mempool.store.get(nextTxKey)
+				// tx may have been removed after it was added to the broadcast queue
+				// thus we just skip over it
+				if wtx != nil {
+					memR.broadcastNewTx(wtx)
+				}
+			}
+		}
+	}()
+	return nil
+}
+
+func (memR *Reactor) OnStop() {
+	// stop all the timers tracking outbound requests
+	memR.requests.Close()
+}
+
+// GetChannels implements Reactor by returning the list of channels for this
+// reactor.
+func (memR *Reactor) GetChannels() []*p2p.ChannelDescriptor {
+	largestTx := make([]byte, memR.opts.MaxTxSize)
+	batchMsg := protomem.Message{
+		Sum: &protomem.Message_Txs{
+			Txs: &protomem.Txs{Txs: [][]byte{largestTx}},
+		},
+	}
+
+	return []*p2p.ChannelDescriptor{
+		{
+			ID:                  mempool.MempoolChannel,
+			Priority:            6,
+			RecvMessageCapacity: batchMsg.Size(),
+			MessageType:         &protomem.Message{},
+		},
+		{
+			ID:                  MempoolStateChannel,
+			Priority:            5,
+			RecvMessageCapacity: maxStateChannelSize,
+			MessageType:         &protomem.Message{},
+		},
+	}
+}
+
+// InitPeer implements Reactor by creating a state for the peer.
+func (memR *Reactor) InitPeer(peer p2p.Peer) p2p.Peer {
+	memR.ids.ReserveForPeer(peer)
+	return peer
+}
+
+// RemovePeer implements Reactor.
+func (memR *Reactor) RemovePeer(peer p2p.Peer, reason interface{}) {
+	peerID := memR.ids.Reclaim(peer.ID())
+	// remove and rerequest all pending outbound requests to that peer since we know
+	// we won't receive any responses from them.
+	outboundRequests := memR.requests.ClearAllRequestsFrom(peerID)
+	for key := range outboundRequests {
+		memR.mempool.metrics.RequestedTxs.Add(1)
+		memR.findNewPeerToRequestTx(key)
+	}
+}
+
+func (memR *Reactor) Receive(chID byte, peer p2p.Peer, msgBytes []byte) {
+	msg := &protomem.Message{}
+	err := proto.Unmarshal(msgBytes, msg)
+	if err != nil {
+		panic(err)
+	}
+	uw, err := msg.Unwrap()
+	if err != nil {
+		panic(err)
+	}
+	memR.ReceiveEnvelope(p2p.Envelope{
+		ChannelID: chID,
+		Src:       peer,
+		Message:   uw,
+	})
+}
+
+// ReceiveEnvelope implements Reactor.
+// It processes one of three messages: Txs, SeenTx, WantTx.
+func (memR *Reactor) ReceiveEnvelope(e p2p.Envelope) {
+	switch msg := e.Message.(type) {
+
+	// A peer has sent us one or more transactions. This could be either because we requested them
+	// or because the peer received a new transaction and is broadcasting it to us.
+	// NOTE: This setup also means that we can support older mempool implementations that simply
+	// flooded the network with transactions.
+	case *protomem.Txs:
```

Comment:
[non-blocking][nit] I'd recommend pulling out the logic in each switch case into its own function so that each path can be more easily unit tested.